Source/WebCore/style/StyleAdjuster.cpp

```diff
@@ -432,6 +432,49 @@ static UnicodeBidi forceBidiIsolationForRuby(UnicodeBidi unicodeBidi)
     return UnicodeBidi::Isolate;
 }
 
+static bool shouldTreatAutoZIndexAsZero(const RenderStyle& style)
+{
+    return style.hasOpacity()
+        || style.hasTransformRelatedProperty()
+        || style.hasMask()
+        || style.clipPath()
+        || style.boxReflect()
+        || style.hasFilter()
+        || style.hasBackdropFilter()
+#if HAVE(CORE_MATERIAL)
+        || style.hasAppleVisualEffect()
+#endif
+        || style.hasBlendMode()
+        || style.hasIsolation()
+        || style.position() == PositionType::Sticky
+        || style.position() == PositionType::Fixed
+        || style.willChangeCreatesStackingContext();
+}
+
+void Adjuster::adjustFromBuilder(RenderStyle& style)
+{
+    // Do some adjustments that don't depend on element or parent style and are safe to cache.
+    // This allows copy-on-write to trigger before caching.
+
+    if (style.hasAutoSpecifiedZIndex()) {
+        if (shouldTreatAutoZIndexAsZero(style))
+            style.setUsedZIndex(0);
+    } else if (style.position() != PositionType::Static)
+        style.setUsedZIndex(style.specifiedZIndex());
+
+    // Cull out any useless layers and also repeat patterns into additional layers.
+    style.adjustBackgroundLayers();
+    style.adjustMaskLayers();
+
+    // Do the same for animations and transitions.
+    style.adjustAnimations();
+    style.adjustTransitions();
+
+    // Do the same for scroll-timeline and view-timeline longhands.
+    style.adjustScrollTimelines();
+    style.adjustViewTimelines();
```

Comment:
This is a bit fragile. How will people know that they need to add to this list in future?

Reply:
The adjustments that can be done here (where no other context is available besides the style itself) should be done here while the rest should be done in the main `Adjuster::adjust()`. I don't think it is particularly fragile as adding things to `adjust()` still works fine but may be non-optimal. 

I only moved things in this patch that were relevant to the case (z-index adjustment) plus the animation things that are obviously not context sensitive.